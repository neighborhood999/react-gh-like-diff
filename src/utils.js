import { unifiedDiff } from 'difflib';
import { format } from 'util';
import { compose } from 'recompose';

const { Diff2Html: { getPrettyHtml } } = require('diff2html');

export const defaultOptions = {
  originalFileName: 'Unknown-File-Name',
  updatedFileName: 'Unknown-File-Name',
  inputFormat: 'diff',
  outputFormat: 'line-by-line',
  showFiles: false,
  matching: 'words',
  matchWordsThreshold: 0.25,
  matchingMaxComparisons: 2500
};

const withErrorMessage = props => {
  const { past, current } = props;

  if (past === undefined || current === undefined) {
    const errorMessage = 'Missing `past` or `current` props!';
    throw errorMessage;
  }

  return props;
};

const gendDiff2Html = ({ past, current, options }) => {
  const nextOptions = Object.assign({}, defaultOptions, options);
  const pastArray = past.split('/\r|\n|\r\n');
  const currentArray = current.split('/\r|\n|\r\n');

  const diffArray = unifiedDiff(pastArray, currentArray, {
    fromFile: nextOptions.originalFileName,
    toFile: nextOptions.updatedFileName
  });

  const diffString = format(
    'diff --git %s %s\n%s',
    nextOptions.originalFileName,
    nextOptions.updatedFileName,
    diffArray.join('\n')
  );

  return getPrettyHtml(diffString, options);
};

export default compose(gendDiff2Html, withErrorMessage);
