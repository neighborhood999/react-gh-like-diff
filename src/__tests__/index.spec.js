import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { readFileSync } from 'fs';
import ReactGhLikeDiff, { RenderDiffResult } from '../index';
import { defaultOptions } from '../utils';

test('should render innerHTML by dangerouslySetInnerHTML', () => {
  const html = '<h1>Hello World</h1>';
  const wrapper = mount(<RenderDiffResult genDiffHTML={html} />);

  expect(wrapper.render().find('h1').text()).toBe('Hello World');
});

test('should render `past` and `current` difference comparison result into HTML', () => {
  const past = 'Hello';
  const current = 'Hello World';
  const options = {
    originalFileName: 'test.txt',
    updatedFileName: 'test.txt'
  };

  const wrapper = mount(
    <ReactGhLikeDiff past={past} current={current} options={options} />
  );

  expect(wrapper.name()).toBe(
    'defaultProps(onlyUpdateForPropTypes(branch(Component)))'
  );

  const tree = renderer
    .create(<ReactGhLikeDiff past={past} current={current} options={options} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test('should render `.diff` file content to HTML', () => {
  const diffFile = readFileSync('./src/__tests__/mock.diff', 'utf8');
  const wrapper = mount(<ReactGhLikeDiff diffString={diffFile} />);

  expect(wrapper.name()).toBe(
    'defaultProps(onlyUpdateForPropTypes(branch(Component)))'
  );

  const tree = renderer
    .create(<ReactGhLikeDiff diffString={diffFile} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
