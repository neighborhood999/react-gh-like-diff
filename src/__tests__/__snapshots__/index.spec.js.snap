// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<div class=\\"d2h-wrapper\\">
        <div id=\\"d2h-906284\\" class=\\"d2h-file-wrapper\\" data-lang=\\"\\">
        <div class=\\"d2h-file-header\\">
        <span class=\\"d2h-file-name-wrapper\\">
        <span class=\\"d2h-icon-wrapper\\"><svg aria-hidden=\\"true\\" class=\\"d2h-icon\\" height=\\"16\\" version=\\"1.1\\" viewBox=\\"0 0 12 16\\" width=\\"12\\">
        <path d=\\"M6 5H2v-1h4v1zM2 8h7v-1H2v1z m0 2h7v-1H2v1z m0 2h7v-1H2v1z m10-7.5v9.5c0 0.55-0.45 1-1 1H1c-0.55 0-1-0.45-1-1V2c0-0.55 0.45-1 1-1h7.5l3.5 3.5z m-1 0.5L8 2H1v12h10V5z\\"></path>
    </svg></span>
        <span class=\\"d2h-file-name\\">test.txt</span>
        <span class=\\"d2h-tag d2h-changed d2h-changed-tag\\">CHANGED</span></span>
        </div>
        <div class=\\"d2h-file-diff\\">
            <div class=\\"d2h-code-wrapper\\">
                <table class=\\"d2h-diff-table\\">
                    <tbody class=\\"d2h-diff-tbody\\">
                    <tr>
        <td class=\\"d2h-code-linenumber d2h-info\\"></td>
        <td class=\\"d2h-info\\">
            <div class=\\"d2h-code-line d2h-info\\">@@ -1 +1 @@</div>
        </td>
    </tr><tr>
        <td class=\\"d2h-code-linenumber d2h-del\\">
          <div class=\\"line-num1\\">1</div>
    <div class=\\"line-num2\\"></div>
        </td>
        <td class=\\"d2h-del\\">
            <div class=\\"d2h-code-line d2h-del\\">
                <span class=\\"d2h-code-line-prefix\\">-</span>
                <span class=\\"d2h-code-line-ctn\\">Hello</span>
            </div>
        </td>
    </tr><tr>
        <td class=\\"d2h-code-linenumber d2h-ins\\">
          <div class=\\"line-num1\\"></div>
    <div class=\\"line-num2\\">1</div>
        </td>
        <td class=\\"d2h-ins\\">
            <div class=\\"d2h-code-line d2h-ins\\">
                <span class=\\"d2h-code-line-prefix\\">+</span>
                <span class=\\"d2h-code-line-ctn\\">Hello<ins> World</ins></span>
            </div>
        </td>
    </tr>
                    </tbody>
                </table>
            </div>
        </div>
    </div>
    </div>",
    }
  }
/>
`;
